### PR TITLE
[PromotionBundle] Update validation process of promotion actions/rules

### PR DESCRIPTION
**Pull request by Wojdylak**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.13
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | N/A
| License         | MIT

### CODE REVIEW

File: config/packages/_sylius.yaml [hunk 1]
```diff
@@ -22,6 +22,51 @@ sylius_payum:
             - 'sylius'
             - 'stripe_checkout'
 
+sylius_promotion:
+    promotion_action:
+        validation_groups:
+            order_percentage_discount:
+                - 'sylius'
+                - 'order_percentage_discount'
```
Comment (diimpp) [inline config/packages/_sylius.yaml]:
```suggestion
                - 'sylius_order_percentage_discount'
```

Still should be prefixed.

Why not keep at at services level at the same place (**tag**) where rest of action/rule configuration exists?

https://github.com/Sylius/Sylius/blob/d3a441f27772712ccc629797303c577a996aab7b/src/Sylius/Bundle/PromotionBundle/Resources/config/services.xml#L38-L40

https://github.com/Sylius/Sylius/blob/d3a441f27772712ccc629797303c577a996aab7b/src/Sylius/Bundle/CoreBundle/Resources/config/services/promotion.xml#L32-L34

https://github.com/Sylius/Sylius/blob/d3a441f27772712ccc629797303c577a996aab7b/src/Sylius/Bundle/CoreBundle/Resources/config/services/promotion.xml#L60-L65

And you will be able to keep validation groups magic to existing DI compiler passes. (For example, add new key `validation-groups` with ability to specify groups directly on tag and write them down for every action/rule)

https://github.com/Sylius/Sylius/blob/1.13/src/Sylius/Bundle/PromotionBundle/DependencyInjection/Compiler/RegisterPromotionActionsPass.php

File: UPGRADE-1.13.md [hunk 2]
```diff
@@ -230,3 +230,43 @@ sylius_payum:
...
+1. A new parameter has been added to specify the validation groups for a given promotion rule.
+   If you have any custom validation groups for your promotion rule, you need to add them to your `config/packages/_sylius.yaml` file.
+   Additionally, if you have your own promotion rule and want to add your validation groups, you can add another key to the `promotion_rule.validation_groups` parameter.
+   This is handled by `Sylius\Bundle\PromotionBundle\Validator\PromotionRuleValidator` and it resolves the groups based on the type of the passed promotion rule.
+
+```yaml
+sylius_promotion:
+    promotion_rule:
+        validation_groups:
+            cart_quantity:
```
Comment (diimpp) [inline UPGRADE-1.13.md]: Just out of curiosity, is it possible to use php const over here?

File: UPGRADE-1.13.md [hunk 3]
```diff
@@ -230,3 +230,43 @@ sylius_payum:
 1. Interface `Sylius\Component\Promotion\Generator\PromotionCouponGeneratorInstructionInterface` has been refactored. It now extends a new interface `Sylius\Component\Promotion\Generator\PromotionCouponGeneratorInstructionReadInterface`, which contains only getter methods.
     - If your services or custom implementations previously relied on `PromotionCouponGeneratorInstructionInterface` for read operations, you should now use `PromotionCouponGeneratorInstructionReadInterface` for better clarity and separation of concerns.
     - This change is backward compatible as long as your implementations or services were using only the getter methods from `PromotionCouponGeneratorInstructionInterface`. However, if you also utilized setter methods, you should continue using `PromotionCouponGeneratorInstructionInterface`.
+
+1. A new parameter has been added to specify the validation groups for a given promotion action.
```
Comment (diimpp) [inline UPGRADE-1.13.md:234]: Feels like it should be added to the docs in current PR as well

File: src/Sylius/Bundle/PromotionBundle/Form/Type/Action/FixedDiscountConfigurationType.php [hunk 4]
```diff
@@ -27,10 +27,6 @@ public function buildForm(FormBuilderInterface $builder, array $options): void
         $builder
             ->add('amount', MoneyType::class, [
                 'label' => 'sylius.form.promotion_action.fixed_discount_configuration.amount',
-                'constraints' => [
```
Comment (diimpp) [inline src/Sylius/Bundle/PromotionBundle/Form/Type/Action/FixedDiscountConfigurationType.php:30]: If somebody made use of `FixedDiscountConfigurationType` the same way `ChannelBasedFixedDiscountConfigurationType` using it, but in their custom promotion action, then validation rules will disappear for them, so I think this is a BC.

There at least should be a mention in UPGRADE. Same for other PromotionBundle form types extended in core.

https://github.com/Sylius/Sylius/blob/8872bafa942392ca93cbd9d99fef14ede79439a4/src/Sylius/Bundle/CoreBundle/Form/Type/Promotion/Action/ChannelBasedFixedDiscountConfigurationType.php#L26-L33

File: src/Sylius/Bundle/CoreBundle/Resources/config/validation/PromotionAction.xml
Comment (diimpp) [inline src/Sylius/Bundle/CoreBundle/Resources/config/validation/PromotionAction.xml:1]:
```xml
                        <constraint name="Type">
                            <option name="groups">
                                <value>order_percentage_discount</value>
                                <value>shipping_percentage_discount</value>
                            </option>
                            <option name="type">numeric</option>
                        </constraint>
```

Usually "meat" of the constraint goes first and groups goes last, otherwise it's hard to read. Would be cool to resort **PromotionAction** and **PromotionRule** files.

```patch
                        <constraint name="Type">
+++                         <option name="type">numeric</option>
                            <option name="groups">
                                <value>order_percentage_discount</value>
                                <value>shipping_percentage_discount</value>
                            </option>
---                         <option name="type">numeric</option>
                        </constraint>
```

File: config/packages/_sylius.yaml (hunk 1 above)
Comment (diimpp) [reply config/packages/_sylius.yaml]: Plus current approach doesn't validate for action/rule **types** and introduces second configuration way for the same thing.

**Review by diimpp — COMMENTED** [review]
Hi, great re-integration of validation rules.

I've left comments in code, but here are key points:
- I think validation groups for action/rules should be moved into already existing action/rule DI tags as `validation-group=['sylius', 'sylius_order_fixed_discount']`.
- Validation groups should be prefixed with `sylius_`
- There is a possible BC with PromotionBundle form validation removal.

File: config/packages/_sylius.yaml (hunk 1 above)
Comment (Wojdylak) [reply config/packages/_sylius.yaml]: This solution is only for the `formType` (UI), but we need something that works for both the UI and the API.

File: UPGRADE-1.13.md (hunk 2 above)
Comment (Wojdylak) [reply UPGRADE-1.13.md]: In the `upgrade` file or the `_sylius` file?

File: UPGRADE-1.13.md (hunk 3 above)
Comment (Wojdylak) [reply UPGRADE-1.13.md:234]: Currently, we are creating new documentation (in gitbook), and this will be included there.

File: src/Sylius/Bundle/PromotionBundle/Form/Type/Action/FixedDiscountConfigurationType.php (hunk 4 above)
Comment (Wojdylak) [reply src/Sylius/Bundle/PromotionBundle/Form/Type/Action/FixedDiscountConfigurationType.php:30]: After discussing with the team, we have determined that this is not a BC, but it is important to add information about this to the upgrade file. I have added it.

File: config/packages/_sylius.yaml (hunk 1 above)
Comment (diimpp) [reply config/packages/_sylius.yaml]: @Wojdylak what do you mean? It's an action registry in the first place, form registry/label are only an additional options, so there is no inherit UI only or API only for this `tag`.

```patch
       $promotionActionRegistry->addMethodCall('register', [$attribute['type'], new Reference($id)]);
       $promotionActionFormTypeRegistry->addMethodCall('add', [$attribute['type'], 'default', $attribute['form_type']]);
+++    $promotionActionsValidationGroups[$attribute['type']] = $attribute['validation_groups'];
```

https://github.com/Sylius/Sylius/blob/05d55a6a2595329fc8b239c2391a44b87e857260/src/Sylius/Bundle/PromotionBundle/DependencyInjection/Compiler/RegisterPromotionActionsPass.php#L38-L40

```patch
        <service id="sylius.promotion_action.fixed_discount" class="Sylius\Component\Core\Promotion\Action\FixedDiscountPromotionActionCommand">
            <argument type="service" id="sylius.proportional_integer_distributor" />
            <argument type="service" id="sylius.promotion.units_promotion_adjustments_applicator" />
            <argument type="service" id="Sylius\Component\Core\Distributor\MinimumPriceDistributorInterface" />
            <argument type="service" id="sylius.currency_converter" />
---         <tag name="sylius.promotion_action" type="order_fixed_discount" label="sylius.form.promotion_action.order_fixed_discount" form-type="Sylius\Bundle\CoreBundle\Form\Type\Promotion\Action\ChannelBasedFixedDiscountConfigurationType" />
+++         <tag name="sylius.promotion_action" type="order_fixed_discount" label="sylius.form.promotion_action.order_fixed_discount" form-type="Sylius\Bundle\CoreBundle\Form\Type\Promotion\Action\ChannelBasedFixedDiscountConfigurationType" validation-groups=["sylius", "sylius_order_fixed_discount"] />
        </service>
```

Comment (Wojdylak) [reply config/packages/_sylius.yaml]: Ok, this looks interesting and would be doable, but
- I can't pass an array through tag attributes
- If I wanted to add a new validation group to an existing action in Sylius, I would have to override the entire definition of service

File: src/Sylius/Bundle/PromotionBundle/Resources/translations/validators.en.yml [hunk 5]
```diff
@@ -37,8 +37,11 @@ sylius:
             min_length: Promotion name must be at least 1 character long.|Promotion name must be at least {{ limit }} characters long.
             not_blank: Please enter promotion name.
     promotion_action:
+        invalid_type: Promotion action type is invalid. Please choose a valid type.
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/PromotionBundle/Resources/translations/validators.en.yml]: Having a list of available types could be handy, same case w/ rules

File: src/Sylius/Bundle/CoreBundle/Resources/config/validation/PromotionRule.xml [hunk 6]
```diff
@@ -13,89 +13,90 @@
...
 <constraint-mapping xmlns="http://symfony.com/schema/dic/constraint-mapping" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://symfony.com/schema/dic/constraint-mapping http://symfony.com/schema/dic/services/constraint-mapping-1.0.xsd">
     <class name="Sylius\Component\Promotion\Model\PromotionRule">
-        <constraint name="Sylius\Bundle\ApiBundle\Validator\Constraints\PromotionRuleGroup">
-            <option name="groups">sylius_api</option>
-        </constraint>
-
         <property name="configuration">
             <constraint name="Collection">
-                <option name="groups">cart_quantity</option>
+                <option name="allowExtraFields">true</option>
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/CoreBundle/Resources/config/validation/PromotionRule.xml:18]: Just to clarify, these are only needed because of a bug in our forms used in the UI.
In those, changing the type does not remove the old configuration and both, new and old, get sent with the request. It'll be addressed in a separate PR since we're not sure of the scope atm

File: src/Sylius/Bundle/CoreBundle/Resources/config/validation/PromotionRule.xml [hunk 7]
```diff
@@ -104,25 +105,27 @@
             </constraint>
 
             <constraint name="Sylius\Bundle\CoreBundle\Validator\Constraints\ChannelCodeCollection">
-                <option name="groups">total_of_items_from_taxon</option>
+                <option name="allowExtraFields">true</option>
+                <option name="groups">sylius_total_of_items_from_taxon</option>
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/CoreBundle/Resources/config/validation/PromotionRule.xml]: Wondering if it wouldn't be better to prefix the group with what the subject is smth like: `sylius_promotion_rule_total_of_items_from_taxon`
We might get a naming overlap with other configurations; not really a problem just might be confusing.

Comment (GSadee) [reply src/Sylius/Bundle/CoreBundle/Resources/config/validation/PromotionRule.xml]: +1 for adding the subject, the name will then be quite long, but will clearly describe what it refers to

File: src/Sylius/Bundle/CoreBundle/Resources/config/validation/PromotionRule.xml (hunk 6 above)
Comment (Wojdylak) [reply src/Sylius/Bundle/CoreBundle/Resources/config/validation/PromotionRule.xml:18]: exactly

**Review by GSadee — DISMISSED** [review]

File: src/Sylius/Bundle/PromotionBundle/DependencyInjection/SyliusPromotionExtension.php [hunk 8]
```diff
@@ -33,6 +33,9 @@ public function load(array $configs, ContainerBuilder $container): void
         $loader->load('services.xml');
         $loader->load(sprintf('services/integrations/%s.xml', $config['driver']));
 
+        $container->setParameter('sylius.promotion.promotion_action.validation_groups', $config['promotion_action']['validation_groups']);
```
Comment (diimpp) [inline src/Sylius/Bundle/PromotionBundle/DependencyInjection/SyliusPromotionExtension.php:36]: Is it possible to check configured actions/rules against action/rules registry, which is configured via compiler pass over here?

File: config/packages/_sylius.yaml (hunk 1 above)
Comment (diimpp) [reply config/packages/_sylius.yaml]: https://github.com/Sylius/Sylius/pull/15628/files#r1426918937

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thanks, Karol! :1st_place_medal:
